### ISSUE TITLE
HostingAbstractionsHostExtensions RunAsync disposes without logging

### ISSUE THREAD

**Issue by snakefoot** [dotnet/runtime#72225]
### Description

The RunAsync in HostingAbstractionsHostExtensions.cs disposes the Host on exit:

https://github.com/dotnet/runtime/blob/2be87c9c976c0cd21b66ea12ae406a559aacaac0/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs#L60-L70

This disposes the ServiceProvider and the LoggerFactory, so unable to perform any logging after this point.

Inside the Host `StartAsync` then it nicely logs `Starting`, but any exception happing in the following code is not caught and logged:

https://github.com/dotnet/runtime/blob/215b39abf947da7a40b0cb137eab4bceb24ad3e3/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L56-L66

And because of the dispose, then any logging is impossible on exception:

### Reproduction Steps

```c#
        var builder = Host.CreateDefaultBuilder();
        builder.ConfigureServices(s => s.AddHostedService<BrokenHostedService>());
        builder.Build().Run();

        class BrokenHostedService : IHostedService
        {
            public Task StartAsync(CancellationToken cancellationToken)
            {
                throw new Exception("Hosted service startup error!");
            }

            public Task StopAsync(CancellationToken cancellationToken)
            {
                return Task.CompletedTask;
            }
        }
```

### Expected behavior

Exception is logged to the Microsoft Console Provider.

### Actual behavior

Application crashes without any trace.

**Comment by eerhardt** [comment]
The application crashes because an unhandled exception is being thrown and killing the Main thread. `StartAsync` is called directly on the Main thread, and until you `await` a Task it continues on the Main thread. See https://github.com/dotnet/runtime/issues/36063 for more info on that.

You can workaround this by wrapping your `StartAsync` method in a `Task.Run`, which will execute your code on a background thread and not block the Main method of the app. However, if you throw an unhandled exception from that background thread, your app won't crash and nothing will be logged unless you log it yourself. See https://github.com/dotnet/runtime/issues/36017. If you want Hosting to handle the logging of unhandled exceptions, you need to inherit from `BackgroundService`.

**Comment by snakefoot** [comment]
Yes I'm aware that with enough effort, then one add enough try-catches everywhere in user-code to ensure exception is logged.

But since having the Host-wrapper, that already now performs logging of Host-"Starting" and Host-"Stopped", then to me it would be natural that it also performed logging of Host-"Starting Failed" on exception. Thus making the user-code cleaner with more focus on the success path, than the error-path.

**Comment by kmcclellan** [comment]
I also find it odd that exceptions thrown starting hosted services aren't logged while [stopping exceptions are](https://github.com/dotnet/runtime/blob/e57438026c25707bf6dd52cd332db657e919bbd4/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L156-L160).

I include this boilerplate in my apps as a workaround. Definitely feels less idiomatic than `host.RunAsync()`.

```c#
await using ((IAsyncDisposable)(host = builder.Build()))
{
    var logger = host.Services.GetService<ILogger<Program>>();

    try
    {
        await host.StartAsync();
    }
    catch (Exception exception)
    {
        logger?.LogCritical(exception, "Failed to start host");
        throw;
    }

    await host.WaitForShutdownAsync();
}
```
